File: web/packages/alloy-audio/src/dsp/svf.ts
// Topology-preserving-transform state variable filter (Zavalishin, "The
// Art of VA Filter Design"). Chosen over biquads because it stays stable
// under audio-rate cutoff modulation — this is the patch TVF, and filter
// envelopes sweep it constantly. Constructed fully open; call setParams to
// shape. Twin: Svf.swift.

export type SvfMode = 'lowpass' | 'bandpass' | 'highpass';

export class Svf {
  private ic1 = 0;
  private ic2 = 0;
  private k = 1;
  private a1 = 0;
  private a2 = 0;
  private a3 = 0;

  constructor(
    private readonly mode: SvfMode,
    private readonly sampleRate: number,
  ) {
    this.setParams(sampleRate * 0.49, 0.707);
  }

  setParams(cutoffHz: number, q: number): void {
    const clamped = Math.min(Math.max(cutoffHz, 10), this.sampleRate * 0.49);
    const g = Math.tan((Math.PI * clamped) / this.sampleRate);
    this.k = 1 / Math.max(q, 0.5);
    this.a1 = 1 / (1 + g * (g + this.k));
    this.a2 = g * this.a1;
    this.a3 = g * this.a2;
  }

  process(x: number): number {
    const v3 = x - this.ic2;
    const v1 = this.a1 * this.ic1 + this.a2 * v3;
    const v2 = this.ic2 + this.a2 * this.ic1 + this.a3 * v3;
    this.ic1 = 2 * v1 - this.ic1;
    this.ic2 = 2 * v2 - this.ic2;
    switch (this.mode) {
      case 'lowpass':
        return v2;
      case 'bandpass':
        return v1;
      case 'highpass':
        return x - this.k * v1 - v2;
    }
  }
}

File: swift/Sources/AlloyAudio/DSP/Svf.swift
import Foundation

/// Topology-preserving-transform state variable filter (Zavalishin).
/// Stable under audio-rate cutoff modulation — this is the patch TVF.
/// Constructed fully open; call setParams to shape.
/// Twin of web src/dsp/svf.ts (canonical).
public enum SvfMode: String, Codable {
    case lowpass
    case bandpass
    case highpass
}

public final class Svf {
    private let mode: SvfMode
    private let sampleRate: Double
    private var ic1 = 0.0
    private var ic2 = 0.0
    private var k = 1.0
    private var a1 = 0.0
    private var a2 = 0.0
    private var a3 = 0.0

    public init(mode: SvfMode, sampleRate: Double) {
        self.mode = mode
        self.sampleRate = sampleRate
        setParams(cutoffHz: sampleRate * 0.49, q: 0.707)
    }

    public func setParams(cutoffHz: Double, q: Double) {
        let clamped = min(max(cutoffHz, 10), sampleRate * 0.49)
        let g = tan(Double.pi * clamped / sampleRate)
        k = 1 / max(q, 0.5)
        a1 = 1 / (1 + g * (g + k))
        a2 = g * a1
        a3 = g * a2
    }

    public func process(_ x: Double) -> Double {
        let v3 = x - ic2
        let v1 = a1 * ic1 + a2 * v3
        let v2 = ic2 + a2 * ic1 + a3 * v3
        ic1 = 2 * v1 - ic1
        ic2 = 2 * v2 - ic2
        switch mode {
        case .lowpass: return v2
        case .bandpass: return v1
        case .highpass: return x - k * v1 - v2
        }
    }
}
